x/programs/cmd/simulator/src/lib.rs

```diff
@@ -392,19 +364,14 @@ where
         self.responses
             .next()
             .ok_or(StepError::Client(ClientError::Eof))?
-    }
-
-    pub fn run_step<T>(
-        &mut self,
-        caller_key: &str,
-        step: &Step,
-    ) -> Result<PlanResponseTyped<T>, StepError>
-    where
-        T: BorshDeserialize,
-    {
-        self._run_step(caller_key, step)?
+            .and_then(|step| {
+                if let Some(err) = step.error {
+                    Err(StepError::Program(err))
+                } else {
+                    Ok(step)
+                }
```

Comment:
You should do this as part of the deserialization so that we can't represent what is effectively an invalid response. We should never have both a response and an error, it should be one or the other.